once/_iterator_wrappers.py

```diff
@@ -24,116 +25,156 @@ def __init__(self) -> None:
+    iterators.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.callable: collections.abc.Callable | None = functools.partial(func, *args, **kwargs)
+        self.generator = self.callable()
+        self.result = IteratorResults()
+        self.generating = False
+
+    def compute_next_action(
+        self, result: IteratorResults, i: int
+    ) -> typing.Tuple[_IteratorAction, typing.Any]:
+        """Must be called with lock."""
+        if i == len(result.items):
+            if result.finished:
+                if result.exception:
+                    raise result.exception
+                return _IteratorAction.RETURNING, None
+            if self.generating:
+                return _IteratorAction.WAITING, None
+            else:
+                # If all of these functions are called with locks, we will never have more than one
+                # caller have GENERATING at any time.
+                self.generating = True
+                return _IteratorAction.GENERATING, None
+        else:
+            return _IteratorAction.YIELDING, result.items[i]
+
+    def successful_completion(self, result: IteratorResults):
+        """Must be called with lock."""
+        result.finished = True
+        self.generating = False
+        self.generator = None  # Allow this to be GCed.
+        self.callable = None  # Allow this to be GCed.
+
+    def record_item(self, result: IteratorResults, item: typing.Any):
+        self.generating = False
+        result.items.append(item)
+
+    def record_exception(self, result: IteratorResults, exception: Exception):
+        """Must be called with lock."""
+        result.finished = True
+        # We need to keep track of the exception so that we can raise it in the same
+        # position every time the iterator is called.
+        result.exception = exception
+        self.generating = False
+        assert self.callable is not None
+        self.generator = self.callable()  # Reset the iterator for the next call.
+        self.result = IteratorResults()
+
+
+class AsyncGeneratorWrapper(_GeneratorWrapperBase):
     """Wrapper around an async generator which only runs once.
 
     Subsequent calls will return results from the first call, which is
     evaluated lazily.
     """
 
+    generator: collections.abc.AsyncGenerator
+
     def __init__(self, func, *args, **kwargs) -> None:
-        self.generator: collections.abc.AsyncGenerator | None = func(*args, **kwargs)
-        self.result = IteratorResults()
-        self.generating = False
+        super().__init__(func, *args, **kwargs)
         self.lock = asyncio.Lock()
 
     async def yield_results(self) -> collections.abc.AsyncGenerator:
-        i = 0
-        send = None
-        next_val = None
-
-        # A copy of self.generating that we can access outside of the lock.
-        generating = None
+        # Fast path for subsequent repeated call:
+        async with self.lock:
```

Comment:
I don't think we should bother with the fast path at all for now, personally. The fast path has one lock and one function call, which is the same as what would happen in the normal path, so the only speedup I would really expect is from using a for loop rather than a while loop, but that's such a micro-optimization that we should have a benchmark to validate that it makes a meaningful enough difference to justify the extra code

Reply:
Removed the fast path and all associated code, left a comment describing why.